README.md

```diff
@@ -0,0 +1,32 @@
+# merkle-cbt-lean
+Merkle tree implementation for constrained memory environments
+
+This crate is based on CBMT implementation in https://crates.io/crates/merkle-cbt
+
+Although indeed efficient both in performance and memory requirements, abovementioned crates memory consumption was found to be too high for its use in really constrained embedded environments (like [Kampela device](https://github.com/Kalapaja/kampela-firmware), or Ledger). We have created this crate to handle proof check with as small memory access and consumption as possible, so that proof lemmas could be stored on external memory or tranferred through serial data interface. Note that conctruction of proof is not optimized, as in typical application it is performed in a regular memory environment and its difficulty is negligible. Root value calculated by this crate and `merkle-cbt` is designed and checked to be identical, this above crate should be used instead of this one where memory limitations are not critical.
+
+We've decided to make this a separate crate to avoid confusion in proof structures and to keep both modulse as minimalistic as they are.
+
+## Concept
+
+To limit memory, we've replaced width graph traverse with depth. Thus, only one branch at a time is stored simultaneously, and CMBT is always much wider than it is tall. To achieve this, we've had to change sorting of the lemmas accordingly, otherwise construction is conceptually same as in abovementioned crate. Main drawback here is roughly quadratic traverse time as opposed to linear width traverse, as, starting from below, we have to traverse whole tree for every leaf.
```

Comment:
```suggestion
To limit memory, we've replaced width graph traverse with depth. Thus, only one branch at a time is stored simultaneously, and CMBT is always much wider than it is tall. To achieve this, we've had to change sorting of the lemmas accordingly, otherwise construction is conceptually same as in above mentioned crate. Main drawback here is roughly quadratic traverse time as opposed to linear width traverse, as, starting from below, we have to traverse whole tree for every leaf.
```